commcare_connect/form_receiver/processor.py

```diff
@@ -303,15 +305,14 @@ def process_deliver_unit(user, xform: XForm, app: CommCareApp, opportunity: Oppo
         user_visit.status = VisitValidationStatus.approved
         user_visit.review_status = VisitReviewStatus.agree
     user_visit.save()
-    if (
-        completed_work is not None
-        and completed_work.completed_count > 0
-        and completed_work.status == CompletedWorkStatus.incomplete
-    ):
-        completed_work.status = CompletedWorkStatus.pending
-        completed_work_needs_save = True
-    if completed_work_needs_save:
-        completed_work.save()
+    if completed_work is not None:
+        if completed_work.completed_count > 0 and completed_work.status == CompletedWorkStatus.incomplete:
+            completed_work.status = CompletedWorkStatus.pending
+            completed_work_needs_save = True
+        if completed_work_needs_save:
+            completed_work.save()
+    with cache.lock(access.id):
```

Comment:
Should the key be more specific than just the ID? For e.g. if we have unrelated locking that also uses `ids` there is possibility for these two have same key.

Reply:
Fixed. [Add lock in update_payment_accrued](https://github.com/dimagi/commcare-connect/pull/416/commits/11d8e4ef9e94252e565bc0e9e6c992081ce01172)